src/picongpu/include/plugins/adios/restart/LoadParticleAttributesFromADIOS.hpp

```diff
@@ -0,0 +1,131 @@
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+
+
+#pragma once
+
+
+#include "types.h"
+#include "simulation_types.hpp"
+#include "plugins/adios/ADIOSWriter.def"
+#include "traits/PICToAdios.hpp"
+#include "traits/GetComponentsType.hpp"
+#include "traits/GetNComponents.hpp"
+#include "traits/Resolve.hpp"
+
+
+namespace picongpu
+{
+
+namespace adios
+{
+using namespace PMacc;
+
+/** Load attribute of a species from ADIOS checkpoint file
+ *
+ * @tparam T_Identifier identifier of species attribute
+ */
+template< typename T_Identifier>
+struct LoadParticleAttributesFromADIOS
+{
+
+    /** read attributes from ADIOS file
+     *
+     * @param params thread params with ADIOS_FILE, ...
+     * @param frame frame with all particles
+     * @param particlePath path to the group in the ADIOS file
+     * @param particlesOffset read offset in the attribute array
+     * @param elements number of elements which should be read the attribute array
+     */
+    template<typename FrameType>
+    HINLINE void operator()(
+                            ThreadParams* params,
+                            FrameType& frame,
+                            const std::string particlePath,
+                            const uint64_t particlesOffset,
+                            const uint64_t elements)
+    {
+
+        typedef T_Identifier Identifier;
+        typedef typename PMacc::traits::Resolve<Identifier>::type::type ValueType;
+        const uint32_t components = GetNComponents<ValueType>::value;
+        typedef typename GetComponentsType<ValueType>::type ComponentType;
+
+        log<picLog::INPUT_OUTPUT > ("ADIOS: ( begin ) load species attribute: %1%") % Identifier::getName();
+
+        const std::string name_lookup[] = {"x", "y", "z"};
+
+        ComponentType* tmpArray = NULL;
+        if( elements > 0 )
+            tmpArray = new ComponentType[elements];
+
+        // dev assert!
+        if( elements > 0 )
+            assert(tmpArray);
+
+        for (uint32_t n = 0; n < components; ++n)
+        {
+            std::stringstream datasetName;
+            datasetName << particlePath << T_Identifier::getName();
+            if (components > 1)
+                datasetName << "/" << name_lookup[n];
+
+            ValueType* dataPtr = frame.getIdentifier(Identifier()).getPointer();
+
+            ADIOS_VARINFO* varInfo = adios_inq_var( params->fp, datasetName.str().c_str() );
+
+            ADIOS_SELECTION* sel = adios_selection_boundingbox( 1, &particlesOffset, &elements );
+            ADIOS_CMD(adios_schedule_read( params->fp,
```

Comment:
there is still an error here, stating `tmpArray` would not contain allocated memory...

```
[1,2]<stdout>:PIConGPUVerbose INPUT_OUTPUT(32) | ADIOS: ( begin ) load species attribut
e: position
[1,0]<stdout>:PIConGPUVerbose INPUT_OUTPUT(32) | ADIOS: error at cmd 'adios_perform_rea
ds( params->fp, 1 )': '-20' Blocking mode at adios_perform_reads() requires that user p
rovides the memory for each read request. Request for variable 8 was scheduled without 
user-allocated memory
```